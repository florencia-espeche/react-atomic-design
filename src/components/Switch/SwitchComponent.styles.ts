import styled, { css } from 'styled-components'

interface StyleProps {
    disabled: boolean
}

export const Wrapper = styled.label<StyleProps>`
    background: ${props =>
        props.disabled
            ? props.theme.colors.neutral.light['300']
            : props.theme.colors.blue.light};

    width: 34px;
    height: 14px;
    border-radius: 24px;

    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;

    span {
        margin-top: 0;
    }
`

export const Circle = styled.span<StyleProps>`
    position: absolute;
    width: 20px;
    height: 20px;
    border-radius: 50%;
    left: 0;

    background: ${props =>
        props.disabled
            ? props.theme.colors.neutral.light['500']
            : props.theme.colors.primary};

    cursor: pointer;
    box-shadow: 2px 2px 4px ${props => props.theme.colors.neutral.light['500']};
    transition: 0.25s;
`

export const Input = styled.input`
    height: 100%;
    width: 100%;
    opacity: 0;
    cursor: pointer;

    &:checked + span {
        left: 50%;
    }

    &:disabled {
        cursor: not-allowed;
    }

    &:disabled + span {
        cursor: not-allowed;
    }
`
