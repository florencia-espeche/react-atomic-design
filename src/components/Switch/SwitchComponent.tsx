import React from 'react'

import { useTheme } from '../../hooks/useTheme'
import { Circle, Input, Wrapper } from './SwitchComponent.styles'
import { SwitchProps } from './SwitchComponent.types'

const SwitchComponent = ({
    onChange,
    checked = false,
    disabled = false
}: SwitchProps) => {
    const theme = useTheme()

    return (
        <Wrapper theme={theme} disabled={disabled}>
            <Input
                type="checkbox"
                value="switch"
                theme={theme}
                disabled={disabled}
                onChange={onChange}
                checked={checked}
            />
            <Circle theme={theme} disabled={disabled} />
        </Wrapper>
    )
}

export default SwitchComponent
