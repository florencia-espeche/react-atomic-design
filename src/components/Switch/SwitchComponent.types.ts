import { ChangeEventHandler } from 'react'

export type SwitchProps = {
    disabled?: boolean
    checked?: boolean
    onChange: ChangeEventHandler<HTMLInputElement>
}
