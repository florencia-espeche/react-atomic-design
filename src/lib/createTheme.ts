import { defaultTheme } from '../styles/defaultTheme'
import { Theme } from '../styles/types'

export const createTheme = <T>(theme: Theme & T) => {
    return {
        ...defaultTheme,
        colors: {
            ...defaultTheme.colors,
            ...theme.colors
        },
        borderRadius: theme.borderRadius || defaultTheme.borderRadius,
        fontTokens: {
            ...defaultTheme.fontTokens,
            ...theme.fontTokens
        },
        fontWeight: theme.fontWeight || defaultTheme.fontWeight,
        name: theme.name || defaultTheme.name
    }
}
