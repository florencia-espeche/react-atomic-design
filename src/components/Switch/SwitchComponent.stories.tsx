import React, { useState } from 'react'

import SwitchComponent from './SwitchComponent'

export default {
    title: 'SwitchComponent',
    component: SwitchComponent
}

export const Default = args => {
    const { checked } = args

    const [isChecked, setIsChecked] = useState(checked)

    const handleChange = () => setIsChecked(!isChecked)

    return (
        <SwitchComponent
            {...args}
            checked={isChecked}
            onChange={handleChange}
        />
    )
}
