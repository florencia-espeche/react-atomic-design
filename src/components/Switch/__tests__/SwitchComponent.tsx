import React from 'react'

import Switch from '../SwitchComponent'
import { render } from '../../../lib/test-utils'

const handleChange = jest.fn()

describe('Switch Component', () => {
    it('Should render a Switch with default styles', () => {
        const { container } = render(<Switch onChange={handleChange} />)

        expect(container.firstChild).toMatchSnapshot()
    })

    it('Should render a checked Switch', () => {
        const { container } = render(
            <Switch checked={true} onChange={handleChange} />
        )

        expect(container.firstChild).toMatchSnapshot()
    })

    it('Should render a disabled Switch', () => {
        const { getByRole, container } = render(
            <Switch disabled={true} onChange={handleChange} />
        )

        const input = getByRole('checkbox')

        expect(input).toHaveStyle('cursor: not-allowed')
        expect(container.firstChild).toMatchSnapshot()
    })
})
