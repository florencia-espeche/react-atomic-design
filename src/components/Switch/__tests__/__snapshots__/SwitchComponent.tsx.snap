// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Component Should render a Switch with default styles 1`] = `
<label
  class="sc-bdfBQB cMDThJ"
>
  <input
    class="sc-dlfnuX fFazWo"
    type="checkbox"
    value="switch"
  />
  <span
    class="sc-gsTEea gAfLgx"
  />
</label>
`;

exports[`Switch Component Should render a checked Switch 1`] = `
<label
  class="sc-bdfBQB cMDThJ"
>
  <input
    checked=""
    class="sc-dlfnuX fFazWo"
    type="checkbox"
    value="switch"
  />
  <span
    class="sc-gsTEea gAfLgx"
  />
</label>
`;

exports[`Switch Component Should render a disabled Switch 1`] = `
<label
  class="sc-bdfBQB hHOdmo"
  disabled=""
>
  <input
    class="sc-dlfnuX fFazWo"
    disabled=""
    type="checkbox"
    value="switch"
  />
  <span
    class="sc-gsTEea bMoSig"
    disabled=""
  />
</label>
`;
